core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala

```diff
@@ -682,7 +682,11 @@ class MasterSuite extends SparkFunSuite
         // an app would be registered with Master once Driver set up
         assert(worker.apps.nonEmpty)
         appId = worker.apps.head._1
-        assert(master.idToApp.contains(appId))
+
+        // we found the case where the test was too fast which all steps were done within
+        // an interval - in this case, we have to check either app is available in master
+        // or marked as completed. See SPARK-30348 for details.
+        assert(master.idToApp.contains(appId) || master.completedApps.exists(_.id == appId))
```

Comment:
I think hit `assert(master.completedApps.exists(_.id == appId))` isn't what we want to see for this test. 

Does a short delay above `assert(master.idToApp.contains(appId))` or increase `MAX_EXECUTOR_RETRIES` works? I'd also prefer not to expose internal data for such a fix.

Reply:
I'd say the test was wrong as I added comment in code; the code is observing the state change on different thread which should be relying on listener stuff; if we can't receive updates via listener, we should take into account that we may not be able to capture the state change sequentially. Regarding adding delay I'll comment again.

Reply:
My concern is, in this test:

1st assert is to make sure that the application exists in `Master.idToApp`

2nd assert is to check the application has been remove from `Master.idToApp` after MAX_EXECUTOR_RETRIES.

But if we find the application has completed in 1st assert, which means it doesn't exist in `Master.idToApp`, I think it has broke original assume.


How about setting up a `CountDownLatch` in `MockExecutorLaunchFailWorker`, and count down before we reach MAX_EXECUTOR_RETRIES ? So we can be sure the application hasn't been removed from `Master.idToApp`.

Reply:
Makes sense, and the suggestion sounds good to me. Let me apply the approach. Thanks!